### PR TITLE
API-6836: Add an asynchronous endpoint - Part 1

### PR DESCRIPTION
**Pull request by johnsgp**

### CODE REVIEW

File: app/uk/gov/hmrc/testppnsmultibox/ppns/BoxIdTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import uk.gov.hmrc.testppnsmultibox.ppns.models.BoxId
+
+case class RequestWithBoxId[A](request: Request[A], boxId: BoxId) extends WrappedRequest[A](request)
+
+@Singleton
+class BoxIdTransformer @Inject() (appConfig: AppConfig, pushPullNotificationConnector: PushPullNotificationConnector)(implicit val executionContext: ExecutionContext)
+    extends ActionTransformer[Request, RequestWithBoxId]
+    with BackendHeaderCarrierProvider {
+
+  private val XClientId = "X-Client-ID"
```
Comment (worthydolt) [inline app/uk/gov/hmrc/testppnsmultibox/ppns/BoxIdTransformer.scala]: Could perhaps be more clearly named. `XClientIdHeaderName` something like that to convey its meaning?

File: app/uk/gov/hmrc/testppnsmultibox/ppns/BoxIdTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  override protected def transform[A](request: Request[A]): Future[RequestWithBoxId[A]] =
+    (for {
+      clientId <- OptionT.fromOption[Future](request.headers.get(XClientId))
+      boxId    <- OptionT.liftF(pushPullNotificationConnector.getBoxId(s"${appConfig.apiContext}##${appConfig.apiVersion}##callbackUrl", clientId)(hc(request)))
+    } yield boxId.getOrElse(
+      throw MissingBoxException(s"A box was not found for clientId $clientId")
+    ))
+      .fold(
+        throw MissingClientIdException(s"$XClientId was not provided as a header")
```
Comment (worthydolt) [inline app/uk/gov/hmrc/testppnsmultibox/ppns/BoxIdTransformer.scala]: Is this string interpolation what you want to do? Isn't this the case where it's missing?

File: app/uk/gov/hmrc/testppnsmultibox/respositories/TimedNotificationRepository.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+import uk.gov.hmrc.mongo.MongoComponent
+import uk.gov.hmrc.mongo.play.json.{Codecs, PlayMongoRepository}
+
+import uk.gov.hmrc.testppnsmultibox.domain.models.TimedNotification
+import uk.gov.hmrc.testppnsmultibox.ppns.models.CorrelationId
+
+@Singleton
+class TimedNotificationRepository @Inject() (mongo: MongoComponent)(implicit val ec: ExecutionContext)
+    extends PlayMongoRepository[TimedNotification](
```
Comment (worthydolt) [inline app/uk/gov/hmrc/testppnsmultibox/respositories/TimedNotificationRepository.scala]: I don't _think_ this repository class defines a TTL index. This is now verboten

File: app/uk/gov/hmrc/testppnsmultibox/ppns/BoxIdTransformer.scala (hunk 2 above)
Comment (johnsgp) [reply app/uk/gov/hmrc/testppnsmultibox/ppns/BoxIdTransformer.scala]: Yes, interpolation is intentional. The error message will be: `X-Client-ID was not provided as a header`

File: app/uk/gov/hmrc/testppnsmultibox/ppns/BoxIdTransformer.scala (hunk 1 above)
Comment (johnsgp) [reply app/uk/gov/hmrc/testppnsmultibox/ppns/BoxIdTransformer.scala]: I'll move it to a separate object named `CustomHeaders`

File: app/uk/gov/hmrc/testppnsmultibox/respositories/TimedNotificationRepository.scala (hunk 3 above)
Comment (johnsgp) [reply app/uk/gov/hmrc/testppnsmultibox/respositories/TimedNotificationRepository.scala]: I'll add an index and an ISpec test

**Review by syed-ali-hmrc — APPROVED** [review]

**Pull request merged**